### ISSUE TITLE
setExpires doesn't work unless set far into the future

### ISSUE THREAD

**Issue by pszabop** [minio/minio-js#545]
setExpires doesn't generate a valid request unless I (as of right now) set the expiration date to be about 6:15 into the future.  I note I'm 8 hours behind UTC.

I note that in setExpires you don't call UTC, you just add a "+Z" at the end.  That could be one source of the problem.

https://github.com/minio/minio-js/blob/18022f0b895e869ebdfc358a488f4ee30db32cf5/src/main/minio.js#L2006

The other weird thing is my default of 1 hour worked at about midnight last night.

I'm using docker containers on OSX and I verified that the date inside the minio container is correct and the date inside the minio-js client is correct.

I haven't looked at S3 in that much detail but I'm curious why  you are using absolute time at all. Shouldn't the time have been relative to whenever the server received the request?

**Comment by pszabop** [comment]
test code (uses blue-tape)

    test('should be able to POST to a presigned URL', function(assert) {

      const filename = 'signedtest.txt';
      var checksum;
 

      var policy = minio.newPostPolicy();

      // Policy restricted only for bucket.
      policy.setBucket(bucket);

      // Policy restricted only for the file object.
      policy.setKey(filename);

      // Policy restricted for incoming objects with keyPrefix.
      // XXX I have no idea WTF this does but it does rename the file set above
      //policy.setKeyStartsWith('keyPrefix');

      // Policy expires in 10 days.
      var expires = new Date;
      //expires.setSeconds(24 * 60 * 60 * 10);
      expires.setSeconds(7 * 60 * 60 - 1 * 50 * 60);
      //expires.setSeconds(6*3600);
      policy.setExpires(expires);
      
      // Only allow 'text'.
      policy.setContentType('text/plain');

      // Only allow content size in range 1B to 1MB.
      policy.setContentLengthRange(1, 1024*1024);


      // create policy (would be done on server)
      return minio.presignedPostPolicyAsync(policy)
      .spread((url, formData) => {
        assert.ok(url.length > 1, 'URL is non empty');
        console.log(formData);
        var req = agent.post(url);
        _.each(formData, function(value, key) {
          req.field(key, value);
        });
        // upload file via the created signed policy
        return req
        .set('Content-Type', 'text/plain')
        .attach('file', 'test/data/test.txt');
      }).then(r => {
        assert.ok(r.ok, 'POST should have returned ok');
        checksum = JSON.parse(r.header.etag);
        return minio.statObjectAsync(bucket, filename);
      }).then(stat => {
        assert.same(stat.etag, checksum, 'after POST checksum matches');

**Comment by pszabop** [comment]
BTW I see that[ Amazon specifies](http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-HTTPPOSTConstructPolicy.html) an absolute UTC time, instead of a relative time to when the request is received by the server.  ARGH, I suppose they architected this 12 years ago but I'd never do use absolute times on distributed systems these days unless I have to...

**Comment by krishnasrinivas** [comment]
@pszabop can you check after this is merged: https://github.com/minio/minio-js/pull/546

**Comment by pszabop** [comment]
400 (Bad Request).  Now that I think about it, shouldn't it have been a 403 (Unauthorized)?.

I shall check your fix.   I don't have a trivial way of patching a docker container outside of NPM so I might wait until it is published or I fix that issue.

**Comment by harshavardhana** [comment]
> I shall check your fix. I don't have a trivial way of patching a docker container outside of NPM so I might wait until it is published or I fix that issue.

You can do 

```
$ npm install --save git+https://github.com/minio/minio-js.git
```

**Comment by pszabop** [comment]
> $ npm install --save git+https://github.com/minio/minio-js.git

Alas I'm a docker newbie.  Just plugging that into my Dockerfile after the normal npm install didn't do the trick.

**Comment by harshavardhana** [comment]
> Alas I'm a docker newbie. Just plugging that into my Dockerfile after the normal npm install didn't do the trick. Now require('minio') can't find minio...

If you share your Dockerfile we can help you. @pszabop

**Comment by pszabop** [comment]
FROM node:latest

    # Create app directory
    RUN mkdir -p /usr/src/app 
    WORKDIR /usr/src/app

    # Install app dependencies
    COPY package.json /usr/src/app/
    RUN npm install
    RUN npm install --save git+https://github.com/minio/minio-js.git
    RUN npm install

    # Bundle app source
    COPY . /usr/src/app/

    EXPOSE 8080
    CMD [ "npm", "test" ]

**Comment by harshavardhana** [comment]
> FROM node:latest
> 
> # Create app directory
> RUN mkdir -p /usr/src/app 
> WORKDIR /usr/src/app
> 
> # Install app dependencies
> COPY package.json /usr/src/app/
> RUN npm install
> RUN npm install --save git+https://github.com/minio/minio-js.git
> RUN npm install

Why is there a need for second 'npm install' ? 'npm install --save' already installs and saves.

**Comment by harshavardhana** [comment]
Fixed and release 3.1.2
